File: R/enkf.R
enkf.analysis <- function(zf, yo, r, loc.inf=1.0){
  k <- nrow(zf)
  ne <- ncol(zf)
  zf.mean <- apply(zf, 1, mean)
  zf.anom <- sweep(zf, 1, zf.mean)
  s <- loc.inf * as.vector(zf.anom %*% zf.anom[k,]) / (ne - 1)
  outer(s, (yo - zf[k,])) / (r + s[k])
}

File: Python/enkf.py
import numpy as np


def analysis(zf, yo, r, loc_inf=1.0):
  k, ne = zf.shape
  zf_mean = zf.mean(axis=1)
  zf_anom = zf - zf_mean[:, None]
  s = loc_inf * (zf_anom @ zf_anom[-1,]) / (ne - 1)
  return np.outer(s, (yo - zf[-1, :])) / (r + s[-1])

